File: ta4j-core/src/main/java/org/ta4j/core/indicators/statistics/VarianceIndicator.java
/*
 * SPDX-License-Identifier: MIT
 */
package org.ta4j.core.indicators.statistics;

import java.util.Objects;

import org.ta4j.core.Indicator;
import org.ta4j.core.indicators.CachedIndicator;
import org.ta4j.core.num.Num;
import org.ta4j.core.num.NumFactory;

/**
 * Variance indicator.
 *
 * <p>
 * The default constructor computes <b>sample variance</b> (divisor
 * {@code n - 1}) for rolling windows. Use {@link #ofPopulation(Indicator, int)}
 * (or the {@link SampleType} constructor) when population variance is required.
 * </p>
 */
public class VarianceIndicator extends CachedIndicator<Num> {

    private final Indicator<Num> indicator;
    private final int barCount;
    private final SampleType sampleType;

    /**
     * Constructor using {@link SampleType#SAMPLE}.
     *
     * @param indicator the indicator
     * @param barCount  the time frame
     * @since 0.22.4
     */
    public VarianceIndicator(Indicator<Num> indicator, int barCount) {
        this(indicator, barCount, SampleType.SAMPLE);
    }

    /**
     * Constructor.
     *
     * @param indicator  the indicator
     * @param barCount   the time frame
     * @param sampleType sample/population variance selection
     * @since 0.22.4
     */
    public VarianceIndicator(Indicator<Num> indicator, int barCount, SampleType sampleType) {
        super(indicator);
        this.indicator = Objects.requireNonNull(indicator, "indicator must not be null");
        this.barCount = Math.max(barCount, 1);
        this.sampleType = Objects.requireNonNull(sampleType, "sampleType must not be null");
    }

    /**
     * Creates an indicator using sample variance ({@code n - 1} divisor).
     *
     * @param indicator the indicator
     * @param barCount  the time frame
     * @return a sample-variance indicator
     * @since 0.22.4
     */
    public static VarianceIndicator ofSample(Indicator<Num> indicator, int barCount) {
        return new VarianceIndicator(indicator, barCount, SampleType.SAMPLE);
    }

    /**
     * Creates an indicator using population variance ({@code n} divisor).
     *
     * @param indicator the indicator
     * @param barCount  the time frame
     * @return a population-variance indicator
     * @since 0.22.4
     */
    public static VarianceIndicator ofPopulation(Indicator<Num> indicator, int barCount) {
        return new VarianceIndicator(indicator, barCount, SampleType.POPULATION);
    }

    @Override
    protected Num calculate(int index) {
        final int startIndex = Math.max(Math.max(0, getBarSeries().getBeginIndex()), index - barCount + 1);
        final int numberOfObservations = index - startIndex + 1;
        final NumFactory numFactory = getBarSeries().numFactory();
        Num variance = numFactory.zero();
        Num average = averageValue(startIndex, index);
        for (int i = startIndex; i <= index; i++) {
            Num pow = indicator.getValue(i).minus(average).pow(2);
            variance = variance.plus(pow);
        }
        final int divisor = sampleType.isSample() ? numberOfObservations - 1 : numberOfObservations;
        if (divisor <= 0) {
            return numFactory.zero();
        }
        return variance.dividedBy(numFactory.numOf(divisor));
    }

    private Num averageValue(int startIndex, int endIndex) {
        Num sum = getBarSeries().numFactory().zero();
        for (int i = startIndex; i <= endIndex; i++) {
            sum = sum.plus(indicator.getValue(i));
        }
        return sum.dividedBy(getBarSeries().numFactory().numOf(endIndex - startIndex + 1));
    }

    @Override
    public int getCountOfUnstableBars() {
        return indicator.getCountOfUnstableBars() + barCount - 1;
    }

    @Override
    public String toString() {
        return getClass().getSimpleName() + " barCount: " + barCount + " sampleType: " + sampleType;
    }
}

File: ta4j-core/src/main/kotlin/org/ta4j/core/indicators/statistics/VarianceIndicator.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.indicators.statistics

import org.ta4j.core.Indicator
import org.ta4j.core.indicators.CachedIndicator
import org.ta4j.core.indicators.SMAIndicator
import org.ta4j.core.num.*
import kotlin.math.max

/**
 * Variance indicator.
 */
class VarianceIndicator(private val indicator: Indicator<Num>, private val barCount: Int) : CachedIndicator<Num>(
    indicator
) {
    private val sma: SMAIndicator

    /**
     * Constructor.
     *
     * @param indicator the indicator
     * @param barCount  the time frame
     */
    init {
        sma = SMAIndicator(indicator, barCount)
    }

    override fun calculate(index: Int): Num {
        val startIndex = max(0, index - barCount + 1)
        val numberOfObservations = index - startIndex + 1
        var variance = numOf(0)
        val average = sma[index]
        for (i in startIndex..index) {
            val pow = (indicator[i] - average).pow(2)
            variance += pow
        }
        variance /= numOf(numberOfObservations)
        return variance
    }

    override fun toString(): String {
        return javaClass.simpleName + " barCount: " + barCount
    }
}